### PR TITLE
Flax support for Stable Diffusion 2

### PR DESCRIPTION
**Pull request by pcuenca**

### CODE REVIEW

File: src/diffusers/models/unet_2d_condition_flax.py [hunk 1]
```diff
@@ -97,11 +97,13 @@ class FlaxUNet2DConditionModel(nn.Module, FlaxModelMixin, ConfigMixin):
         "DownBlock2D",
     )
     up_block_types: Tuple[str] = ("UpBlock2D", "CrossAttnUpBlock2D", "CrossAttnUpBlock2D", "CrossAttnUpBlock2D")
+    only_cross_attention: Union[bool, Tuple[bool]] = False,
```
Comment (camenduru) [inline src/diffusers/models/unet_2d_condition_flax.py:100]: little , syntax error

File: src/diffusers/models/unet_2d_blocks_flax.py [hunk 2]
```diff
@@ -68,6 +70,8 @@ def setup(self):
                 n_heads=self.attn_num_head_channels,
                 d_head=self.out_channels // self.attn_num_head_channels,
```
Comment (camenduru) [inline src/diffusers/models/unet_2d_blocks_flax.py:71]: TypeError: unsupported operand type(s) for //: 'int' and 'tuple'

**Review by camenduru — CHANGES_REQUESTED** [review]

File: src/diffusers/models/unet_2d_condition_flax.py [hunk 3]
```diff
@@ -97,11 +97,13 @@ class FlaxUNet2DConditionModel(nn.Module, FlaxModelMixin, ConfigMixin):
         "DownBlock2D",
     )
     up_block_types: Tuple[str] = ("UpBlock2D", "CrossAttnUpBlock2D", "CrossAttnUpBlock2D", "CrossAttnUpBlock2D")
+    only_cross_attention: Union[bool, Tuple[bool]] = False
     block_out_channels: Tuple[int] = (320, 640, 1280, 1280)
     layers_per_block: int = 2
-    attention_head_dim: int = 8
+    attention_head_dim: Union[int, Tuple[int]] = 8,
```
Comment (camenduru) [inline src/diffusers/models/unet_2d_condition_flax.py:103]: little , jumped here 😋

**Comment by camenduru** [comment]
thanks @pcuenca ♥ it is working 🎉

**Comment by camenduru** [comment]
I converted to flax I put it https://huggingface.co/flax/stable-diffusion-2 if someone needs

File: src/diffusers/models/attention_flax.py [hunk 4]
```diff
@@ -104,17 +103,22 @@ class FlaxBasicTransformerBlock(nn.Module):
             Hidden states dimension inside each head
         dropout (:obj:`float`, *optional*, defaults to 0.0):
             Dropout rate
+        cross_attention_dim (`int`, *optional*):
```
Comment (patrickvonplaten) [inline src/diffusers/models/attention_flax.py:106]:
```suggestion
```

File: src/diffusers/models/attention_flax.py [hunk 5]
```diff
@@ -104,17 +103,22 @@ class FlaxBasicTransformerBlock(nn.Module):
...
+            The size of the context vector for cross attention.
```
Comment (patrickvonplaten) [inline src/diffusers/models/attention_flax.py:107]:
```suggestion
```

**Comment by patrickvonplaten** [comment]
Super cool!

Should we maybe add one slow test for SD-2?

File: src/diffusers/models/attention_flax.py (hunk 4 above)
Comment (pcuenca) [reply src/diffusers/models/attention_flax.py:106]: Thanks!

**Comment by pcuenca** [comment]
> Should we maybe add one slow test for SD-2?

I added a couple of integration tests to ensure that the output from the Flax UNet is close enough to the output from PyTorch. This works within a tolerance of `1e-2` in my hardware, despite PyTorch values being generated in `float16` and Flax ones in `bfloat16`. This sounds like the UNet implementation appears to be correct.

The tests may fail on the real testing hardware though (V100 and TPU). I'll adapt if that's the case.

File: tests/models/test_models_unet_2d_flax.py [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+        hidden_states = jnp.array(load_hf_numpy(self.get_file_format(seed, shape)), dtype=dtype)
+        return hidden_states
+
+    @parameterized.expand(
+        [
+            # fmt: off
+            [83, 4, [-0.2323, -0.1304, 0.0813, -0.3093, -0.0919, -0.1571, -0.1125, -0.5806]],
+            [17, 0.55, [-0.0831, -0.2443, 0.0901, -0.0919, 0.3396, 0.0103, -0.3743, 0.0701]],
+            [8, 0.89, [-0.4863, 0.0859, 0.0875, -0.1658, 0.9199, -0.0114, 0.4839, 0.4639]],
+            [3, 1000, [-0.5649, 0.2402, -0.5518, 0.1248, 1.1328, -0.2443, -0.0325, -1.0078]],
```
Comment (pcuenca) [inline tests/models/test_models_unet_2d_flax.py:51]: These are the same slices from the torch `float16` GPU test. JAX results are within reasonable tolerance on the same GPU, despite being done in `bfloat16`.

In this case the use of `parameterized` is slow as each test triggers a compilation. If we want to make it faster we should replace them with a single test and a loop.

File: tests/models/test_models_unet_2d.py [hunk 7]
```diff
@@ -639,3 +639,29 @@ def test_compvis_sd_inpaint_fp16(self, seed, timestep, expected_slice):
...
+            [83, 4, [0.1514, 0.0807, 0.1624, 0.1016, -0.1896, 0.0263, 0.0677, 0.2310]],
+            [17, 0.55, [0.1164, -0.0216, 0.0170, 0.1589, -0.3120, 0.1005, -0.0581, -0.1458]],
+            [8, 0.89, [-0.1758, -0.0169, 0.1004, -0.1411, 0.1312, 0.1103, -0.1996, 0.2139]],
+            [3, 1000, [0.1214, 0.0352, -0.0731, -0.1562, -0.0994, -0.0906, -0.2340, -0.0539]],
+            # fmt: on
+        ]
+    )
+    @require_torch_gpu
+    def test_stabilityai_sd_v2_fp16(self, seed, timestep, expected_slice):
+        model = self.get_unet_model(model_id="stabilityai/stable-diffusion-2", fp16=True)
```
Comment (patrickvonplaten) [inline tests/models/test_models_unet_2d.py:655]: Thanks a lot for adding this!

File: tests/models/test_models_unet_2d_flax.py (hunk 6 above)
Comment (patrickvonplaten) [reply tests/models/test_models_unet_2d_flax.py:51]: Cool!

**Review by patrickvonplaten — APPROVED** [review]
Super nice PR! Thanks for adding all those tests :-)

**Comment by pcuenca** [comment]
Merging then to deploy the backend. Thanks!

**Pull request merged**